### DISCUSSION TITLE
Virtual sensor to get weather temperature?

### DISCUSSION THREAD

**Discussion by wojtulab** [emsesp/EMS-ESP32#2174] [Q&A]
**Subject:** Integrating a Virtual Sensor for Outdoor Temperature to Influence Heating Curve

Hello everyone,

I'm exploring the possibility of integrating a virtual sensor into my system that could read the current outdoor temperature for my specific region (via an online weather API or similar source) and use this data to adjust the heating curve accordingly. 

The idea is to have this virtual sensor act as an alternative to a physical outdoor temperature sensor, feeding real-time temperature data to the system so it can influence the heating curve dynamically.

Has anyone implemented a setup like this or could advise on the feasibility of such an integration? I'd appreciate any guidance on possible solutions, recommended APIs, or integration methods that could help achieve this.

Thank you!

**Comment by MichaelDvP** [comment]
There is a buderus outdoor wireless sensor T1RF. We know that it uses the wireless base at device-Id 0x50 to send just the 0xD1 telegram with outdoor temperature. I don't know if the base needs some extra communication, maybe only sending ther temperature is enough. 
Try to send a raw telegram every minute with command `system send "50 00 D1 00 xx xx` where xx xx is the hex temperature in 0.1°C resolution, eg. 12.5°C => 125 => 0x00D7 gives `system send "50 00 D1 00 00 D7`
For testing you can use the scheduler with a fixed temperature value:
![grafik](https://github.com/user-attachments/assets/b322385f-ba3d-4db7-aa46-f6cbba352d04)

Goto logger and select log all to record the telegrams and see if a device (thermostat/boiler) responses to this telegram or repeats the value.

**Comment by wojtulab** [comment]
@MichaelDvP  should I enable outdoor temperature sensor before that?
![image](https://github.com/user-attachments/assets/74e91268-6fe3-45f5-80f3-b23e949e4837)

![image](https://github.com/user-attachments/assets/f9256493-e3a0-48bd-83ff-8993b1a76978)



`2024-11-05 13:59:06.006 T 321: [emsesp] me(0x0B) -R-> thermostat(0x10), RC300OutdoorTemp(0x023A), length: 25
2024-11-05 13:59:06.070 T 322: [emsesp] thermostat(0x10) -W-> me(0x0B), RC300OutdoorTemp(0x023A), data: FF 9C`

![image](https://github.com/user-attachments/assets/d01c59b3-4a72-4c86-ba4f-2f684179cf59)

**Reply by MichaelDvP** [reply]
The timer should send every minute, not every 13:59 hours.

**Reply by wojtulab** [reply]
there is no way to set it every minute:
![image](https://github.com/user-attachments/assets/324997ac-3f0e-4cf3-973f-e9a8d8de6fdf)

**Comment by wojtulab** [comment]
@MichaelDvP could you tell me what next ?

**Comment by wojtulab** [comment]
I created cron script at remote device:
`curl -X POST http://192.168.1.45/api/system/send      -H "Authorization: Bearer "   -H "Content-Type: application/json"      -d '{"cmd": "send", "data": "50 00 D1 00 00 D7"}'`

![image](https://github.com/user-attachments/assets/2daecfd3-700e-4e98-bb4d-56e5dcc30b6f)

system send "50 00 D1 00 xx xx where xx xx is the hex temperature in 0.1°C

I thought that temperature is here:
TELEGRAM="50 00 D1 ${TEMP_HEX} 00 D7"

**Reply by MichaelDvP** [reply]
If you like to send the real temperature you have to replace 00 D7 by the temperature-value HEX(highbyte(temperatur*10)) HEX(lowbyte(temperature*10). 
But for first test i think you can use a fixed value, maybe send more often (15 sec) and see if the thermostat shows this value as outdoor temperature.

BTW: you can log the telegrams and click download, safe it to file an drag/drop the file here. This is much better readable that screenshots with only a few lines.

**Comment by wojtulab** [comment]
I wrote script
```
TEMP_DEC=$1
TEMP_HEX=$(printf '%02X' "$TEMP_DEC")
TELEGRAM="50 00 D1 00 ${TEMP_HEX}"

# sending telegram via cURL
curl -X POST http://$EMS_ESP_IP/api/system/send \
     -H "Authorization: Bearer $BEARER_TOKEN" \
     -H "Content-Type: application/json" \
     -d "{\"cmd\": \"send\", \"data\": \"$TELEGRAM\"}"
```

**Comment by wojtulab** [comment]
there is no temperature available at rc300 or api output
[log.txt](https://github.com/user-attachments/files/17664360/log.txt)

**Reply by MichaelDvP** [reply]
Ok, than it is much more complicated to emulate this sensor. 
I'll try to make a emulation with register device 0x50 including version repyl and send temperature periodical. Will take a bit.
My system in ems (not ems+) and has a physical sensor connected to boiler, so can't test the virtual. 
If the test is ready i will inform you for testing.

**Reply by MichaelDvP** [reply]
Please try this build: https://github.com/MichaelDvP/EMS-ESP32/releases/tag/test

You should find in settings application a checkbox for Emulate outdoorsensor:
![Screenshot 2024-11-09 at 15-08-04 EMS-ESP](https://github.com/user-attachments/assets/1fdaf2b7-7b63-403f-a691-41d48e718f88)
If enabled there should be Wireless sensorbase detected in devices with outdoortemp 0°C
![Screenshot 2024-11-09 at 15-08-39 EMS-ESP](https://github.com/user-attachments/assets/14e36d34-c318-4ed3-b0e7-0c9500718e13)

Edit it to the actual outdoortemp, or use `connect/outdoortemp` as api/mqtt/telent command.
Then try if your thermostat detect the temperature right and allows to configure to outdoor based control.

Make a log of the communication to see if thermostat tries to get more values from this device.

**Comment by wojtulab** [comment]
It's working but temperature appers only for 2-3 seconds after that I have 0*C. 
![image](https://github.com/user-attachments/assets/600be3a0-6b93-4775-8352-6afa2728ced2)

rc310 thermostat:
![image](https://github.com/user-attachments/assets/133d5c90-8bae-44c7-9436-05fbbcad94c6)


logs
[log-1.txt](https://github.com/user-attachments/files/17700124/log-1.txt)

**Reply by MichaelDvP** [reply]
You have to disable the system send command, this is now a bus-collision. Please make the log a bit longer (~5 minutes).

**Reply by wojtulab** [reply]
I rewrited it to use "outdoortemp/connect" instead of system/send but I've got wrong temperature:

```
{
    "outdoortemp": 50
}
```

I sent temp: 50 00 D1 00 00 14 = 2*C // Ok, changed to decimal instead HEX...

**Reply by MichaelDvP** [reply]
You have to send `connect/outdoortemp` with a single value e.g. 9.7, but this is not needed to repeat frequently, you use this command only on change of external temp.  For testing you can just use the web interface and click on the pencil.

But looking on #1584 with much communication to the sensorbase it's maybe much more needed to emulate the sensor. We have to try step by step.

**Reply by wojtulab** [reply]
It's working
`{
    "outdoortemp": 2
}`

**Comment by wojtulab** [comment]
I can't see outdoor temperature at thermostat - should be visible at the screen
![image](https://github.com/user-attachments/assets/b2b88297-d28f-4ac7-835d-61df947ae4fc)

but the temp curve is ok
![image](https://github.com/user-attachments/assets/11e76b03-9167-4c0e-964e-d280311358d5)

**Reply by MichaelDvP** [reply]
> I can't see outdoor temperature at thermostat - should be visible at the screen

I think you have to configure it in the thermostat  standarddisplay settings.

> but the temp curve is ok

That's a step forward. Is it now possible to set the controlmode of the thermostat heatingcircuit to outdoor (weather compensated)?

**Reply by wojtulab** [reply]
but at Buderus I can see new flashing icon: 
![Screenshot_20241111_133723_Gallery](https://github.com/user-attachments/assets/cf275f2d-a98c-4119-9392-a7cd6b22dd76)

**Reply by wojtulab** [reply]
controlmode of the thermostat heatingcircuit set to outdoor (weather compensated) 
but 
I can't find at thermostat standarddisplay settings to view outdoor temperature

**Reply by MichaelDvP** [reply]
You have not send the system info, so i don't know what thermostat you have. 
The RC300 manual says there is in general settings a menu standard display and you can choose, different temperatures (solar, outdoor, etc . to show in the upper right corner. Check the manual from your thermostat.

**Reply by wojtulab** [reply]
sorry it's rc310 and yes, I found it. It's already set to outside but temperature is still missing.
![20241111_150957](https://github.com/user-attachments/assets/f565f771-b1ce-4e53-8c92-6c383a79c6bc)
![20241111_150952](https://github.com/user-attachments/assets/73588e93-ccd7-4814-a034-7f4347092e60)
![20241111_151004](https://github.com/user-attachments/assets/d689e26b-eac3-4063-af2a-b20a4de44344)

**Comment by wojtulab** [comment]
@MichaelDvP log file:
[log (1).txt](https://github.com/user-attachments/files/17701560/log.1.txt)

**Reply by MichaelDvP** [reply]
Please disable the outdoorsensor, don't send temperature to connect/outdoortemp. Wait 15 Minutes.
Than start logging, goto settings and enable outdoorsensor, immediate go back to logger page, wait wait 5 minutes and download the log. 
There should be some communication from thermostat to the 0x50 device, if know what the thermostat wants to know, we maybe can send it.

**Reply by wojtulab** [reply]
ok, how to set log level?

**Reply by MichaelDvP** [reply]
Loglevel all, want to see the telegrams from first detect of device 0x50.

**Reply by wojtulab** [reply]
ok, done
[log-5min.txt](https://github.com/user-attachments/files/17706387/log-5min.txt)

**Reply by wojtulab** [reply]
> Loglevel all, want to see the telegrams from first detect of device 0x50.

`2024-11-11 18:19:09.752 TRACE 2946: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:19:24.816 TRACE 3020: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:19:39.819 TRACE 3028: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:19:54.991 TRACE 3067: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:20:10.206 TRACE 3073: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:20:25.308 TRACE 3145: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:20:40.614 TRACE 3151: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:20:55.778 TRACE 3176: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:21:10.954 TRACE 3183: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:21:26.248 TRACE 3256: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:21:41.304 TRACE 3261: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:21:56.588 TRACE 3286: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:22:12.591 TRACE 3296: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:22:27.634 TRACE 3364: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:22:43.680 TRACE 3382: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:22:58.846 TRACE 3394: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:23:13.961 TRACE 3404: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:23:29.089 TRACE 3472: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:23:44.146 TRACE 3490: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:23:59.311 TRACE 3502: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
2024-11-11 18:24:14.426 TRACE 3512: [emsesp] connect(0x50) -> all(0x00), B, RFOutdoorTemp(0xD1), data: 00 0A 00
`

**Reply by MichaelDvP** [reply]
Hmm, i've expected some telegrams with thermostat -> connect and type 0xF7, 0x02, .. But there is nothing.
Strange that the graph works, but not the display. So thermostat recognize the temperature in one case, but not in the other, Does the thermostat allow to switch the heating circuit to outdoor?
In #1584 there is a newer wireless controller that sends out telegrams 0x880, 0x889, 0x87F, but i don't know what the values mean. 

Can you set loglevel all, then restart emsesp, go to logger after restart and download the log after 5 minutes. Also try to send once
(maybe in terminal), `call system send "00 00 F7 00 D1 00 00 03"` and `call system send "00 10 F7 00 D1 00 00 03"`

**Reply by wojtulab** [reply]
> Does the thermostat allow to switch the heating circuit to outdoor?

yes, set to outdoor 

Yes I also noticed that I can see graph of outside temperature but can't see actual measured outside temperature

[log3.txt](https://github.com/user-attachments/files/17714910/log3.txt)

**Reply by Sbried** [reply]
>  In #1584 there is a newer wireless controller that sends out telegrams 0x880, 0x889, 0x87F, but i don't know what the values mean.
> 
Regarding the telegrams in question (as far as I know):
0x87F: wireless info
- Offset 0 = Battery Status in 0,5% steps
- Offset 1 = Error status

0x880: Outdoor sensor status
- Offset 0, Bit 0 = wireless connected
- Offset 0, Bit 1 = wireless paired
- Offset 1 = pairing status (0=invalid, 4=paired; 1-3=?)

0x889: RSSI values
- Offset 0 = Device-ID
- Offset 1 = RSSI max
- Offset 2 = RSSI min

**Reply by wojtulab** [reply]
> Can you set loglevel all, then restart emsesp, go to logger after restart and download the log after 5 minutes. Also try to send once (maybe in terminal), `call system send "00 00 F7 00 D1 00 00 03"` and `call system send "00 10 F7 00 D1 00 00 03"`

so what will be next?

**Reply by MichaelDvP** [reply]
> so what will be next?

I've changed the wireless base to product_id 218, add the T1RF, and send also the connected/paired signal 0x880. 
Maybe this helps, please try my 3.7.1-test.6 build. https://github.com/MichaelDvP/EMS-ESP32/releases/tag/test
Post a log and the system info from emsesp-help page.

**Reply by wojtulab** [reply]
the same scenario - can't view actual temperature at thermostat, as you can see T1RF is 0, and wireless base is 1 (available)
![image](https://github.com/user-attachments/assets/c89c67d5-a1d1-4c42-87bd-6f726392dca2)

logfile:
[log (2).txt](https://github.com/user-attachments/files/17749183/log.2.txt)

**Reply by wojtulab** [reply]
@MichaelDvP any idea what next?

**Reply by MichaelDvP** [reply]
I have compared the logs and (so far) did not find any differences. The thermostat calculates the damped outdoortemp correct and is set to room+footpoint control. So he gets the outdoortemp correctly. Don't know why you can't show it on the main screen. Does other settings work?
I've changed the test build back to simple wireless base without extra T1RF.

**Reply by wojtulab** [reply]
> Does other settings work?

actually, I can't set to controlmode of the thermostat heatingcircuit to outdoor.

It seems that the outdoor temperature is available, but the sensor itself is not recognized by the thermostat. Therefore, I cannot set the boiler to operate based on the outdoor sensor.

**Reply by MichaelDvP** [reply]
>  I can't set to controlmode of the thermostat heatingcircuit to outdoor.

Oh, sad. I'm out of ideas. 
Can you log and restart the thermostat/heating, maybe we miss a query to the wireless base.

**Reply by MichaelDvP** [reply]
Have you ever tested with the F7 commands i've mentioned? 
As maybe better variant: With outdoorsensor enabled send `call system send "50 10 F7 00 D1 00 00 03"`
(or try `call system send "50 00 F7 00 D1 00 00 03"`)

**Reply by wojtulab** [reply]
> Have you ever tested with the F7 commands i've mentioned?

yes, without success :/

**Comment by oliof** [comment]
Here is another startup log from an RM200 and a T1 RF sensor. I power cycled the heating system at 15:43.

[log(5).txt](https://github.com/user-attachments/files/18320858/log.5.txt)

**Comment by oliof** [comment]
My heating system has a hardwired outside sensor it is using, but EMS-ESP 3.7.1 shows the measured temp of the T1 RF continuously. It's currently living in the basement and does report basement temps since the system doesn't have enough range for me to place outside right now. From the ems-esp console / `show values`:

```
Connect Module: RFM 200/M200, RFM200 (DeviceID:0x50, ProductID:218, Version:34.02) (1)
  outside temperature (outdoortemp): 22.2 °C
```

If I disconnect the RM200 and/or the T1 RF, I get an error A1009 (remote thermostat missing) which I can't seem to clear without reconnecting it. Any hints on that would be welcome (-:

**Comment by 2beeRockz** [comment]
I also own a RC200RF, T1RF and RFM200, and would like to replace the T1RF with a virtual version due to missing reliability of the outdoor sensor... is there anything i can provide you with in order to get this virtual T1RF functional?

**Reply by MichaelDvP** [reply]
Difficult. Both RF devices are managed by the RMF200, I guess emulating only one on the bus is not possible.
But if you can make a long log from starting the heating system, we can learn more about the startup interview of these devcies.
Do you have any settings to configure the the RF devices in menu of the heating system, a  log of changing these settings would also be helpfull. Adding the system info with devices and the received telegrams are also helpfull.

**Reply by VlastiBroucek** [reply]
I would be very careful with virtual sensors - Bureau of Meteorology here in Australia has been showing outside temperature for my area up to 15 degrees lower than it actually was for the past 8 hours or so. At one point it was showing 25 degrees (-15 instead of 10) less than reality...

Would not want to rely on that! 😁🤣😂

**Comment by wieland-arthur** [comment]
Any chance that the .bin firmware gets published or the virtual outdoor sensor is added to the dev branch?
